examples/generative/text_generator.exs

```diff
@@ -0,0 +1,122 @@
+# Based on https://machinelearningmastery.com/text-generation-lstm-recurrent-neural-networks-python-keras/
+Mix.install([
+  {:axon, github: "elixir-nx/axon"},
+  {:nx, "~> 0.2.1"},
+  {:exla, "~> 0.2.2"},
+  {:req, "~> 0.3.0"}
+])
+
+EXLA.set_as_nx_default([:tpu, :cuda, :rocm, :host])
+
+defmodule TextGenerator do
+  require Axon
+
+  @download_url "https://www.gutenberg.org/files/11/11-0.txt"
+  @filename "alice_in_wonderland.txt"
+  @sequence_length 100
+  @batch_size 128
+
+  def book_text do
+    unless File.exists?(@filename) do
+      Req.get!(@download_url)
+      |> then(fn %{body: content} -> File.write!(@filename, content) end)
+    end
+
+    File.read!(@filename)
+  end
+
+  def build_model(characters_count) do
+    Axon.input({nil, @sequence_length, 1}, "input_chars")
+    |> Axon.lstm(256)
+    |> then(fn {_, out} -> out end)
+    |> Axon.nx(fn x -> Nx.slice_along_axis(x, @sequence_length - 1, 1, axis: -2) end)
+    |> Axon.dropout(rate: 0.2)
+    |> Axon.flatten()
+    |> Axon.dense(characters_count, activation: :softmax)
+  end
+
+  def generate(model, params, init_seq, char_to_int, characters) do
+    init_seq =
+      init_seq
+      |> String.downcase()
+      |> String.to_charlist()
+      |> Enum.map(&Map.fetch!(char_to_int, &1))
+
+    Enum.reduce(1..80, init_seq, fn _, seq ->
+      init_seq =
+        seq
+        |> Enum.take(-@sequence_length)
+        |> Nx.tensor()
+        |> Nx.divide(Enum.count(characters))
+        |> Nx.reshape({1, @sequence_length, 1})
+
+      char =
+        Axon.predict(model, params, init_seq)
+        |> Nx.argmax()
+        |> Nx.to_number()
+
+      seq ++ [char]
+    end)
+    |> Enum.map(&Enum.at(characters, &1))
+  end
+
+  def transform_text(text, char_to_int, characters) do
+    train_data =
+      text
+      |> Enum.map(&Map.fetch!(char_to_int, &1))
+      |> Enum.chunk_every(@sequence_length, 1, :discard)
+      # We don't want the last chunk since we don't have a prediction for it.
+      |> Enum.slice(0..-2)
+      |> Nx.tensor()
+      |> Nx.divide(Enum.count(characters))
+      |> Nx.reshape({:auto, @sequence_length, 1})
+      |> Nx.to_batched_list(@batch_size)
+
+    train_labels =
+      text
+      |> Enum.drop(@sequence_length)
+      |> Enum.map(&Map.fetch!(char_to_int, &1))
+      |> Nx.tensor()
+      |> Nx.reshape({:auto, 1})
+      |> Nx.equal(Nx.tensor(Enum.to_list(0..(Enum.count(characters) - 1))))
+      |> Nx.to_batched_list(@batch_size)
+
+    {train_data, train_labels}
+  end
+
+  def run do
+    normalized_book_text =
+      book_text()
+      |> String.downcase()
+      |> String.replace(~r/[^A-Za-z \.\n]/, "")
+      |> String.to_charlist()
+
+    # Extract all then unique characters we have. Optionally we can sort them.
+    characters = normalized_book_text |> Enum.uniq() |> Enum.sort()
+    # Create a mapping for every character
+    char_to_int = Enum.with_index(characters) |> Enum.into(%{})
```

Comment:
It may be confusing that characters are already integers, perhaps if we name it `char_to_idx` it will be more clear?